File: framework/lib/grid_row.dart
part of uxml;

/**
 * Implements a grid row.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class GridRow extends GridDef {
  /**
   * MinHeight Property Definition.
   */
  static PropertyDefinition minHeightProperty;

  /**
   * MaxHeight Property Definition.
   */
  static PropertyDefinition maxHeightProperty;

  /**
   * Height Property Definition.
   */
  static PropertyDefinition heightProperty;

  static ElementDef gridrowElementDef;

  GridRow() : super() {
  }

  /**
   * Gets or sets height of a grid row.
   */
  num get height => length;

  set height(num value) {
    setProperty(heightProperty, value);
  }

  /**
   * Gets or sets the minimum height of a grid row.
   */
  num get minHeight => minLength;

  set minHeight(num value) {
    setProperty(minHeightProperty, value);
  }

  /**
   * Gets or sets the maximum height of a grid row.
   */
  num get maxHeight => maxLength;

  set maxHeight(num value) {
    setProperty(maxHeightProperty, value);
  }

  static void _heightChangeHandler(UxmlElement target,
                                   PropertyDefinition property,
                                   Object oldValue,
                                   Object newValue) {
    GridRow row = target;
    if (newValue == PropertyDefaults.NO_DEFAULT) {
      row.lengthDefined = false;
    } else {
      row.length = newValue;
    }
    if (row.owner != null) {
      row.owner._gridDefChanged();
    }
  }

  static void _maxHeightChangeHandler(UxmlElement target,
                                      PropertyDefinition property,
                                      Object oldValue,
                                      Object newValue) {
    GridRow row = target;
    if (newValue == PropertyDefaults.NO_DEFAULT) {
      row.maxLengthDefined = false;
    } else {
      row.maxLength = newValue;
    }
    if (row.owner != null) {
      row.owner._gridDefChanged();
    }
  }

  static void _minHeightChangeHandler(UxmlElement target,
                                      PropertyDefinition property,
                                      Object oldValue,
                                      Object newValue) {
    GridRow row = target;
    row.minLength = (newValue == PropertyDefaults.NO_DEFAULT) ? 0 : newValue;
    if (row.owner != null) {
      row.owner._gridDefChanged();
    }
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => gridrowElementDef;

  /** Registers component. */
  static void registerGridRow() {
    minHeightProperty = ElementRegistry.registerProperty("minHeight",
        PropertyType.NUMBER, PropertyFlags.NONE, _minHeightChangeHandler, 0.0);
    maxHeightProperty = ElementRegistry.registerProperty("maxHeight",
        PropertyType.NUMBER, PropertyFlags.NONE, _maxHeightChangeHandler, 0.0);
    heightProperty = ElementRegistry.registerProperty("height",
        PropertyType.NUMBER, PropertyFlags.NONE, _heightChangeHandler, 0.0);
    gridrowElementDef = ElementRegistry.register("GridRow",
        UxmlElement.baseElementDef,
        [minHeightProperty, maxHeightProperty, heightProperty], null);
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/GridRow.java
package com.hello.uxml.tools.framework;

import java.util.EnumSet;

/**
 * Implements a grid row.
 *
 * @author ferhat
 */
public class GridRow extends GridDef {
  /** Height Property Definition */
  public static PropertyDefinition heightPropDef = PropertySystem.register("Height", Double.class,
      GridRow.class,
      new PropertyData(Double.NaN, EnumSet.of(PropertyFlags.None), new PropertyChangeListener() {
        @Override
        public void propertyChanged(PropertyChangedEvent e) {
          ((GridRow) e.getSource()).length = ((Double) e.getNewValue()).doubleValue();
        }}));

  /** MinHeight Property Definition */
  public static PropertyDefinition minHeightPropDef = PropertySystem.register("MinHeight",
      Double.class, GridRow.class,
      new PropertyData(Double.NaN, EnumSet.of(PropertyFlags.None), new PropertyChangeListener() {
        @Override
        public void propertyChanged(PropertyChangedEvent e) {
          ((GridRow) e.getSource()).minLength = ((Double) e.getNewValue()).doubleValue();
        }}));

  /** MaxHeight Property Definition */
  public static PropertyDefinition maxHeightPropDef = PropertySystem.register("MaxHeight",
      Double.class, GridRow.class,
      new PropertyData(Double.NaN, EnumSet.of(PropertyFlags.None), new PropertyChangeListener() {
        @Override
        public void propertyChanged(PropertyChangedEvent e) {
          ((GridRow) e.getSource()).maxLength = ((Double) e.getNewValue()).doubleValue();
        }}));

  /**
   * Sets/returns height property.
   */
  public void setHeight(double value) {
    setProperty(heightPropDef, value);
  }

  public double getHeight() {
    return length;
  }

  /**
   * Sets/returns minimum height property.
   */
  public void setMinHeight(double value) {
    setProperty(minHeightPropDef, value);
  }

  public double getMinHeight() {
    return minLength;
  }

  /**
   * Sets/returns maximum height property.
   */
  public void setMaxHeight(double value) {
    setProperty(maxHeightPropDef, value);
  }

  public double getMaxHeight() {
    return maxLength;
  }
}
